File: 2015/ruby/day23/part2.rb
#!/usr/bin/ruby

begin
    File.open('input.txt', 'r') do |file|
        instructions = []
        registers = {'a' => 1, 'b' => 0}
        index = 0

        file.each_line do |line|
            instructions.push(line.tr(',', '').chomp)
        end

        while index < instructions.count
            instruction = instructions[index].split

            case instruction[0]
            when 'hlf'
                registers[instruction[1]] /= 2
            when 'tpl'
                registers[instruction[1]] *= 3
            when 'inc'
                registers[instruction[1]] += 1
            when 'jmp'
                index += instruction[1].to_i
                next
            when 'jie'
                if registers[instruction[1]].even?
                    index += instruction[2].to_i - 1
                end
            when 'jio'
                if registers[instruction[1]] == 1
                    index += instruction[2].to_i - 1
                end
            end

            index += 1
        end

        puts registers['b']
    end
rescue SystemCallError
    puts 'input.txt not found.'
end

File: 2016/python/day12/part2.py
#!/usr/bin/python


def parse(value, registers):
    return registers[value] if value in registers else int(value)


with open('input.txt') as file_handler:
    instructions = [line.rstrip().split() for line in file_handler]

registers = {'a': 0, 'b': 0, 'c': 1, 'd': 0}
index = 0

while index < len(instructions):
    instr = instructions[index]
    index += 1

    if instr[0] == 'cpy':
        registers[instr[2]] = parse(instr[1], registers)
    elif instr[0] == 'inc':
        registers[instr[1]] += 1
    elif instr[0] == 'dec':
        registers[instr[1]] -= 1
    elif parse(instr[1], registers) != 0:
        index += int(instr[2]) - 1

print registers['a']
